### PR TITLE
Add option to specify git clone options like --depth for Zephyr repo clone via init

### PR DESCRIPTION
**Pull request by usmanimtiaz63**
Although `clone-depth` is provided to use for `update` command, the `--depth` option is not provided for `init` without `-l`. It adds this option.

### CODE REVIEW

File: src/west/app/project.py [hunk 1]
```diff
@@ -186,6 +186,9 @@ def do_add_parser(self, parser_adder):
         parser.add_argument('-m', '--manifest-url',
                             help='''manifest repository URL to clone;
                             cannot be combined with -l''')
+        parser.add_argument('-d', '--depth',
+                            help='''clone depth of repository;
+                            cannot be combined with -l''')
```
Comment (pdgendt) [inline src/west/app/project.py:191]: I think we don't want to split all possible git options like this, instead we could do the same as the `update -o` command:

```python
        parser.add_argument('-o', '--clone-opt', action='append', default=[],
                            help='''additional option to pass to 'git clone'
                            (e.g. 'o=--depth=1');
                            may be given more than once;
                            cannot be combined with -l''')
```

**Review by pdgendt — COMMENTED** [review]
Squash related changes into a single commit, and add tests.

**Comment by pdgendt** [comment]
My previous comment is still valid, squash all commits into a single commit, and force push to this PR.

Add test case(s) to `tests/test_project.py`

File: src/west/app/project.py [hunk 2]
```diff
@@ -177,7 +177,7 @@ def do_add_parser(self, parser_adder):
             parser_adder,
             usage='''
 
-  %(prog)s [-m URL] [--mr REVISION] [--mf FILE] [directory]
+  %(prog)s [-m URL] [--mr REVISION] [--mf FILE] [directory] [-o=MORE_GIT_OPTIONS]
```
Comment (pdgendt) [inline src/west/app/project.py:180]: Nit
```suggestion
  %(prog)s [-m URL] [--mr REVISION] [--mf FILE] [-o=GIT_CLONE_OPTION] [directory]
```

**Comment by marc-hb** [comment]
At the risk of stating the obvious, the  slightly slower alternative is to `git clone manifest-repo.git` and `west init -l`. So this PR is not adding any "new feature" but just saving one step.

**Comment by marc-hb** [comment]
> My previous comment is still valid, squash all commits into a single commit, and force push to this PR.

@usmanimtiaz63 , the Zephyr project does not use GitHub in the usual way. It uses it in the "traditional git", pre-GitHub way. Long story in:

https://github.com/zephyrproject-rtos/zephyr/issues/39194#issuecomment-1022428673

**Comment by mbolivar** [comment]
> why @mbolivar didn't add this back sooner

All the optimization work we did on west update was targeted at CI use cases where the manifest repository was already cloned, since the git forge workflows typically give you an environment where an MR is already checked out.

This looks fine to me!

File: src/west/app/project.py [hunk 3]
```diff
@@ -186,6 +186,10 @@ def do_add_parser(self, parser_adder):
         parser.add_argument('-m', '--manifest-url',
                             help='''manifest repository URL to clone;
                             cannot be combined with -l''')
+        parser.add_argument('-o', '--clone-opt', action='append', default=[],
+                            help='''additional option to pass to 'git clone'
+                            (e.g. '-o=--depth=1'); may be given more than once;
```
Comment (marc-hb) [inline src/west/app/project.py:191]: Just FYI, I tested this PR and compared:
```
1. west init -o=--depth=1
2. west init -o=--filter=tree:0
```

The latter took 30% longer and 10% more disk but has a complete `git log` and does not break `git describe`. We've been using it in CI for a couple years now and it's a much better choice there.

**Review by marc-hb — COMMENTED** [review]
Minor test change requested, everything else looks good, thanks!

**Comment by marc-hb** [comment]
> Local Test run:

Just FYI: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#use-copy-paste

File: tests/test_project.py [hunk 4]
```diff
@@ -1647,6 +1647,28 @@ def test_init_local_with_empty_path(repos_tmpdir):
...
+        cmd(['init', '-l', '-o=--depth=1', west_tmpdir])
+
+
+def test_init_with_clone_option_depth_one(repos_tmpdir):
+    # Test that 'west init -o=--depth=1' only clones depth 1
+
+    west_tmpdir = repos_tmpdir / 'workspace'
+
+    cmd(['init', '-o=--depth=1', west_tmpdir])
+    west_tmpdir.chdir()
```
[the remark this answers is not in the document]
Comment (usmanimtiaz63) [reply tests/test_project.py]: Removed.

File: tests/test_project.py [hunk 5]
```diff
@@ -1647,6 +1647,28 @@ def test_init_local_with_empty_path(repos_tmpdir):
...
+    assert subprocess.check_output([GIT, 'log', '--skip=0'],
+                                   cwd=west_tmpdir / 'zephyr') != b''
+    assert subprocess.check_output([GIT, 'log', '--skip=1'],
+                                   cwd=west_tmpdir / 'zephyr') == b''
```
[the remark this answers is not in the document]
Comment (usmanimtiaz63) [reply tests/test_project.py]: Updated

[the remark this answers is not in the document]
Comment (usmanimtiaz63) [reply tests/test_project.py]: $ tox  -- -k test_init_with_clone_option_depth_one
.pkg: _optional_hooks> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py3: install_package> python -I -m pip install --force-reinstall --no-deps /home/usman/workdir/sls/zephyrproject/west/.tox/.tmp/package/42/west-1.2.99.tar.gz
py3: commands[0]> python -m pytest --cov-report=html --cov=west -k test_init_with_clone_option_depth_one '--basetemp=/home/usman/workdir/sls/zephyrproject/west/.tox/py3/tmp/pytest with space/'
========================================== test session starts ==========================================
platform linux -- Python 3.11.2, pytest-8.3.3, pluggy-1.5.0
cachedir: .tox/py3/.pytest_cache
rootdir: /home/usman/workdir/sls/zephyrproject/west
configfile: pyproject.toml
plugins: cov-5.0.0
collected 200 items / 199 deselected / 1 selected                                                       

tests/test_project.py .                                                                           [100%]

---------- coverage: platform linux, python 3.11.2-final-0 -----------
Coverage HTML written to dir htmlcov


============================= 1 passed, 199 deselected in 113.73s (0:01:53) =============================
py3: commands[1]> python -m flake8 --config=/home/usman/workdir/sls/zephyrproject/west/tox.ini /home/usman/workdir/sls/zephyrproject/west
py3: commands[2]> python -m mypy --config-file=/home/usman/workdir/sls/zephyrproject/west/tox.ini --package=west
src/west/app/project.py:1005: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 12 source files
  py3: OK (119.67=setup[4.62]+cmd[114.02,0.54,0.49] seconds)
  congratulations :) (119.70 seconds)

File: tests/test_project.py [hunk 6]
```diff
@@ -1647,6 +1647,25 @@ def test_init_local_with_empty_path(repos_tmpdir):
...
+        cmd(['init', '-l', '-o=--depth=1', west_tmpdir])
+
+
+def test_init_with_clone_option_depth_one(repos_tmpdir):
+    # Test that 'west init -o=--depth=1' only clones depth 1
+
+    west_tmpdir = repos_tmpdir / 'workspace'
+
+    cmd(['init', '-o=--depth=1', west_tmpdir])
+    assert 1 == int(subprocess.check_output([GIT, 'rev-list', '--count', '--all'],
```
Comment (marc-hb) [inline tests/test_project.py:1665]: `--all` does not make sense to me, it assumes `--single-branch` is the default so please replace it with `HEAD`. The fewer assumptions the better.

`--max-count=5` is a small and cheap optimization.

```suggestion
    assert 1 == int(subprocess.check_output([GIT, 'rev-list', '--count', '--max-count=5', 'HEAD'],
```

Comment (usmanimtiaz63) [reply tests/test_project.py]:
```sh
tox  -- -k test_init_with_clone_option_depth_one
.pkg: _optional_hooks> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /home/usman/workdir/sls/zephyrproject/.venv/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py3: install_package> python -I -m pip install --force-reinstall --no-deps /home/usman/workdir/sls/zephyrproject/west/.tox/.tmp/package/48/west-1.2.99.tar.gz
py3: commands[0]> python -m pytest --cov-report=html --cov=west -k test_init_with_clone_option_depth_one '--basetemp=/home/usman/workdir/sls/zephyrproject/west/.tox/py3/tmp/pytest with space/'
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.11.2, pytest-8.3.3, pluggy-1.5.0
cachedir: .tox/py3/.pytest_cache
rootdir: /home/usman/workdir/sls/zephyrproject/west
configfile: pyproject.toml
plugins: cov-5.0.0
collected 200 items / 199 deselected / 1 selected                                                                                                                                                                 

tests/test_project.py .                                                                                                                                                                                     [100%]

---------- coverage: platform linux, python 3.11.2-final-0 -----------
Coverage HTML written to dir htmlcov


================================================================================== 1 passed, 199 deselected in 79.76s (0:01:19) ===================================================================================
py3: commands[1]> python -m flake8 --config=/home/usman/workdir/sls/zephyrproject/west/tox.ini /home/usman/workdir/sls/zephyrproject/west
py3: commands[2]> python -m mypy --config-file=/home/usman/workdir/sls/zephyrproject/west/tox.ini --package=west
src/west/app/project.py:1005: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 12 source files
  py3: OK (84.19=setup[3.07]+cmd[80.00,0.51,0.62] seconds)
  congratulations :) (84.23 seconds)
```

**Review by pdgendt — APPROVED** [review]

**Review by marc-hb — APPROVED** [review]

File: tests/test_project.py [hunk 7]
```diff
@@ -1647,6 +1647,25 @@ def test_init_local_with_empty_path(repos_tmpdir):
...
+    with pytest.raises(subprocess.CalledProcessError):
+        cmd(['init', '-l', '-o=--depth=1', west_tmpdir])
+
+
+def test_init_with_clone_option_depth_one(repos_tmpdir):
+    # Test that 'west init -o=--depth=1' only clones depth 1
+
+    west_tmpdir = repos_tmpdir / 'workspace'
+
+    cmd(['init', '-o=--depth=1', west_tmpdir])
```
Comment (marc-hb) [inline tests/test_project.py:1664]: @usmanimtiaz63 it looks like you have a fast enough network connection! See fix in:
- #855

**Pull request merged**